### PR TITLE
Fix gradient propagation for in-place array assignments

### PR DESCRIPTION
**Pull request by shivansh023023**
## Description

This PR fixes a bug where gradient propagation for in-place array assignments on vector and matrix types (e.g., `arr[i] = wp.vec3(1.0, 2.0, 3.0)` and `arr[i] += wp.vec3(1.0, 2.0, 3.0)`) would fail to build correct `wp.Tape` nodes and backpropagate gradients. 

The bug lied in the fact that intermediate vector/matrix values inside arrays default to [Reference](cci:2://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/codegen.py:686:0-688:36) types during code generation. By explicitly stripping the [Reference](cci:2://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/codegen.py:686:0-688:36) type inside [vector_assign_dispatch_func](cci:1://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/builtins.py:9806:0-9837:37), [vector_assign_copy_value_func](cci:1://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/builtins.py:9873:0-9875:19), [matrix_assign_dispatch_func](cci:1://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/builtins.py:10081:0-10167:37), and [matrix_assign_copy_value_func](cci:1://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/builtins.py:10195:0-10197:19), we cleanly register the backwards assignment node for autodifferentiation. Additionally, [emit_AugAssign](cci:1://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/codegen.py:3476:4-3661:18) in [codegen.py](cci:7://file:///home/shivanshsingh/Desktop/warp/warp/warp/_src/codegen.py:0:0-0:0) now correctly redirects augmented array assignments to the standard base assignment protocol. 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://nvidia.github.io/warp/user_guide/contribution_guide.html).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes. *(N/A)*

## Test plan

- Uncommented the existing `TODO` gradient propagation assertions for array assignments in [warp/tests/test_vec.py](cci:7://file:///home/shivanshsingh/Desktop/warp/warp/warp/tests/test_vec.py:0:0-0:0) and [warp/tests/matrix/test_mat.py](cci:7://file:///home/shivanshsingh/Desktop/warp/warp/warp/tests/matrix/test_mat.py:0:0-0:0).
- Ran the entire test suite via `python warp/tests/test_vec.py` and `python warp/tests/matrix/test_mat.py` successfully. 
- Created and validated custom reproduction scripts mapping `x[0] = y[0]` and `x[0][0] += 5.0` with gradients, confirming that `tape.backward(loss=x)` accurately sets `y.grad[0]` to `1.0`.

## Bug fix

```python
import warp as wp
import numpy as np
wp.init()

@wp.kernel
def assign_kernel(x: wp.array(dtype=float), y: wp.array(dtype=float)):
    x[0] = y[0] # In-place array assignment

@wp.kernel
def aug_assign_kernel(x: wp.array(dtype=wp.vec3)):
    x[0][0] += 5.0 # In-place vector assignment

y = wp.array([10.0], dtype=float, requires_grad=True)
x = wp.array([0.0], dtype=float, requires_grad=True)
vec = wp.zeros(1, dtype=wp.vec3, requires_grad=True)

tape = wp.Tape()
with tape:
    wp.launch(assign_kernel, dim=1, inputs=[x, y])
    wp.launch(aug_assign_kernel, dim=1, inputs=[vec])

tape.backward(loss=x)
# Previously, y.grad[0] would be 0.0 or trigger a Codegen KeyError
# With this PR, it properly propagates the gradient: 1.0
print(f"y.grad: {y.grad.numpy()}")


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * More robust vector/matrix assignment and indexing: container types are normalized, partial vs. full indexing for multi-dimensional assignments is handled correctly, and augmented-assignment/subscript behavior improved; kernel error messages shortened.

* **Tests**
  * Re-enabled and updated array-assignment tests, including restored gradient assertion and an exposed kernel for matrix array assignment.

* **Documentation**
  * Added Apache 2.0 license headers and formatting cleanups.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by shi-eric** [comment]
Thanks for working on this. The core bug this PR targets has since been fixed on `main` by `560ad813` using a different slot-level lowering for array-rooted composite component writes. That implementation avoids the older whole-element `assign_copy` / `array_store` reconstruction path used here, which was incomplete for some composite shapes and conflicted with the current codegen structure.

The existing vector and matrix component assignment regressions now pass on current `main`, and this PR no longer applies cleanly against the current codegen path.

There are still adjacent gaps around array-rooted composite augmented assignment and matrix row/slice writes. I filed those separately as #1451 with focused reproductions so they can be handled independently.

Given that this PR is now stale and superseded by the merged implementation, I’m going to close it in favor of #1451.
